lib/matplotlib/axes/_axes.py

```diff
@@ -5153,79 +5176,39 @@ def fill_between(self, x, y1, y2=0, where=None, interpolate=False,
                                  'is not 1-dimensional.')
 
+        pad_size = x.size
+        if step == 'between':
+            pad_size -= 1
+
         if where is None:
-            where = True
+            where = np.ones(pad_size).astype(bool)
         else:
             where = np.asarray(where, dtype=bool)
-            if where.size != x.size:
+            if where.size != x.size and step != 'between':
                 cbook.warn_deprecated(
                     "3.2",
                     message="The parameter where must have the same size as x "
                             "in fill_between(). This will become an error in "
                             "future versions of Matplotlib.")
-        where = where & ~functools.reduce(np.logical_or,
-                                          map(np.ma.getmask, [x, y1, y2]))
-
-        x, y1, y2 = np.broadcast_arrays(np.atleast_1d(x), y1, y2)
-
-        polys = []
-        for ind0, ind1 in cbook.contiguous_regions(where):
-            xslice = x[ind0:ind1]
-            y1slice = y1[ind0:ind1]
-            y2slice = y2[ind0:ind1]
-            if step is not None:
-                step_func = cbook.STEP_LOOKUP_MAP["steps-" + step]
-                xslice, y1slice, y2slice = step_func(xslice, y1slice, y2slice)
-
-            if not len(xslice):
-                continue
-
-            N = len(xslice)
-            X = np.zeros((2 * N + 2, 2), float)
-
-            if interpolate:
-                def get_interp_point(ind):
-                    im1 = max(ind - 1, 0)
-                    x_values = x[im1:ind + 1]
-                    diff_values = y1[im1:ind + 1] - y2[im1:ind + 1]
-                    y1_values = y1[im1:ind + 1]
-
-                    if len(diff_values) == 2:
-                        if np.ma.is_masked(diff_values[1]):
-                            return x[im1], y1[im1]
-                        elif np.ma.is_masked(diff_values[0]):
-                            return x[ind], y1[ind]
-
-                    diff_order = diff_values.argsort()
-                    diff_root_x = np.interp(
-                        0, diff_values[diff_order], x_values[diff_order])
-                    x_order = x_values.argsort()
-                    diff_root_y = np.interp(diff_root_x, x_values[x_order],
-                                            y1_values[x_order])
-                    return diff_root_x, diff_root_y
-
-                start = get_interp_point(ind0)
-                end = get_interp_point(ind1)
-            else:
-                # the purpose of the next two lines is for when y2 is a
-                # scalar like 0 and we want the fill to go all the way
-                # down to 0 even if none of the y1 sample points do
-                start = xslice[0], y2slice[0]
-                end = xslice[-1], y2slice[-1]
 
-            X[0] = start
-            X[N + 1] = end
+        # Broadcast scalar values
+        y1 = np.broadcast_to(y1, pad_size, subok=True)
```

Comment:
You've abstracted a lot of the logic, but still have logic outside your helper that is repeated here and in `fill_betweenx`.   Is there a good reason for that?